teps/0138-decouple-api-and-feature-versioning.md

```diff
@@ -72,8 +91,330 @@ Per [TEP-0033](https://github.com/tektoncd/community/blob/main/teps/0033-tekton-
+
+| Feature stability level | Default  |
+| ----------------------- | -------- |
+| Stable                  | Enabled; Cannot be disabled  |
+| Beta                    | Disabled |
+| Alpha                   | Disabled |
+
+Note that per feature flags that have stabilized cannot be disabled. We will deprecate the per feature flag after it has become stable and then remove it eventually after the deprecation period according to the API compatibility policy. Notification of the deprecation and the removal of the per feature flags will be given to users via release notes. Cluster operators who do not want such opt-in features would have enough notice to implement admission controllers on their own to disable the feature.
+For example, when a new future feature `pipeline-in-pipeline` becomes stable in v0.55, it would be enabled by default and cannot be disabled after the release. We would need to include in the release note of v0.55 that we are enabling the `pipeline-in-pipeline` feature by default and deprecating its feature flag. And after the deprecation period, we would remove the feature flag.
+
+The behaviour of existing `enable-api-fields` flag with per feature flag:
+- Any current beta features can be enabled with `enable-api-fields` set to “beta” or “alpha”.
+- Any current alpha features can be enabled with `enable-api-fields` set to “alpha”. When current alpha features are promoted to beta, they can be enabled with `enable-api-fields` set to `beta` or `alpha`.
+- The existing features are not adopting this new mechanism of having one flag per feature. It will not be possible to enable existing features using per-feature flags.
+  - We cannot enable existing features using per-feature flags because this would not be backwards compatible. If we allowed this, the individual flag would have precedence over the grouped flag, which means that to preserve backwards compatibility, the individual flag would need to be on by default for beta features and off by default for alpha features. However, this would not be backwards compatible for cluster operators who set `enable-api-fields` to `stable`, since they would also need to override the new `beta` level per feature flags. 
+
+- **Cluster Operators:** For new features, cluster operators will explicitly turn on or off each features in the ConfigMap. They will be able to choose to turn on a single feature. See future work for how cluster operators would communicate with their users on the list of features enabled.
+
+- **Task and Pipeline Authors:** Tekton Pipeline and Task authors will get to know the list of features that are turned on from their service providers. See [future work](#future-work) for the better communication from cluster operators to users for more details.
+
+### Sunset `enable-api-fields` after existing features stabilize
```

Comment:
I agree that we shall give heads up for existing features. But I am not sure if keeping it in the current TEP would be helpful since basically all new features coming in after this TEP being merged shall use per-feature flag model ideally.(Plus merging this TEP sooner so to avoid such situations) Maybe giving heads up in the WG could be a nice option WDYT?

Reply:
Do you mean that new features coming in should follow per-feature flag after this TEP is `merged` or `implemented`?

Reply:
Yes:) And it looks like most maintainers are on board with this TEP, we are hoping to get this merged by EOD. I will go ahead and update the API-compatibility policy ASAP.